app/controllers/api/v1/packages_controller.rb

```diff
@@ -207,6 +208,19 @@ def designate
         send_stock_item_response
       end
 
+      api :PUT, "/v1/packages/1/actions/:action_name", "Executes an action on a package"
+      param_group :operations
+
+      def exec_action
+        Package::Operations.perform_action(@package,
```

Comment:
@swatijadhav This `perform_action` method would accept basically any action here, which means it's basically opening up the inventory to any writes, which is way too permissive.

For example, `dispatch`, `inventory`, `move` actions are reserved to some unique operations, they all come with their own validation/conditions. So those should not be allowed in the inventory unless we go through the proper entry points.

I would suggested locking down the entrypoint to something which represents `"pure"` quantity changes such as a `loss/recycled/...`. There's already `register_loss/register_gain` entry points which include the proper validation.

I've actually added some code in my quantity-related PR which could be useful

```
    ##
    # Registers either a gain or a loss action depending on the change value
    #
    # @param [Package] package the package that went missing
    # @param [Integer] quantity the quantity that was lost
    # @param [Location|Id] location the location to add the quantity to (or its id)
    #
    def register_quantity_change(package, delta:, location:)
      return if delta.zero?
      delta.positive? ?
        register_gain(package, quantity: delta, to_location: location) :
        register_loss(package, quantity: delta, from_location: location)
    end
```

Reply:
@patrixr 

In `stock_operations` concern, for every action check will be performed for valid incremental or decremental actions. refer https://github.com/crossroads/api.goodcity/pull/939/files#diff-1ec071393023822c7c83f0bae29e140aR72

Currently only added check for decremental actions. (later will add a check for other incremental actions) and for quantity, we are accepting only positive values and those will be updated further based on action type (either positive or negative type, I think you have already added changes for it, which am reusing.)

Should we add changes in that implementation and pass the `-/+` quantity value from the params?

Reply:
@swatijadhav Yes, that is a good check for `loss/gain`, `trash/untras`, ...

But we still shouldnt allow `move`, `dispatch`, `inventory`.

The fact that some actions are only decremental or only incremental has nothing to do with the meaning of the action. e.g, `uninventory` is a decremental action, which would pass through those validation, except it should **never** be allowed outside of its own entry point because of all of the business logic associated with that action :)

Reply:
Agree, will update the check at https://github.com/crossroads/api.goodcity/pull/939/files#diff-1ec071393023822c7c83f0bae29e140aR72 and will only include the actions which will be allowed for quantity change and not other actions.

Reply:
Note, just in case it isn't clear, the `quantity` param is always a delta, meaning the amount that changed. And not the resulting quantity

Reply:
Thanks :)
Maybe we can define those "pure" actions somewhere. Meaning the simple ones that don't have any business logic attached to them. Can't think of a good name though 🤔 

Note, all `gain actions` *must* check the following : 

`raise Goodcity::NotInventorizedError unless PackagesInventory.inventorized?(package)`